exec/java-exec/src/test/java/org/apache/drill/test/QueryBuilder.java

```diff
@@ -0,0 +1,314 @@
+    Preconditions.checkNotNull(queryText, "Query not provided.");
+    client.client().runQuery(queryType, queryText, listener);
+  }
+
+  /**
+   * Run the query, return an easy-to-use event listener to process
+   * the query results. Use when the result set is large. The listener
+   * allows the caller to iterate over results in the test thread.
+   * (The listener implements a producer-consumer model to hide the
+   * details of Drill listeners.)
+   *
+   * @return the query event listener
+   */
+
+  public BufferingQueryEventListener withEventListener( ) {
+    BufferingQueryEventListener listener = new BufferingQueryEventListener( );
+    withListener(listener);
+    return listener;
+  }
+
+  public long printCsv() {
+    return print(Format.CSV);
+  }
+
+  public long print( Format format ) {
+    return print(format,20);
+  }
+
+  public long print(Format format, int colWidth) {
+    return runAndWait( new PrintingResultsListener( client.cluster().config( ), format, colWidth ) );
+  }
+
+  /**
+   * Run a query and optionally print the output in TSV format.
+   * Similar to {@link QueryTestUtil#test} with one query. Output is printed
+   * only if the tests are running as verbose.
+   *
+   * @return the number of rows returned
+   */
+  public long print() {
+    DrillConfig config = client.cluster().config( );
+
+    // Note: verbose check disabled until that change is
+    // committed.
+
+    boolean verbose = ! config.getBoolean(QueryTestUtil.TEST_QUERY_PRINTING_SILENT) /* ||
+                      DrillTest.verbose() */;
+    if (verbose) {
+      return print(Format.TSV, VectorUtil.DEFAULT_COLUMN_WIDTH);
+    } else {
+      return run().recordCount();
+    }
+  }
+
+  public long runAndWait(UserResultsListener listener) {
+    AwaitableUserResultsListener resultListener =
+        new AwaitableUserResultsListener(listener);
+    withListener( resultListener );
+    try {
+      return resultListener.await();
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Submit an "EXPLAIN" statement, and return text form of the
+   * plan.
+   * @throws Exception if the query fails
+   */
+
+  public String explainText() throws Exception {
+    return explain(ClusterFixture.EXPLAIN_PLAN_TEXT);
+  }
+
+  /**
+   * Submit an "EXPLAIN" statement, and return the JSON form of the
+   * plan.
+   * @throws Exception if the query fails
+   */
+
+  public String explainJson() throws Exception {
+    return explain(ClusterFixture.EXPLAIN_PLAN_JSON);
+  }
+
+  public String explain(String format) throws Exception {
+    queryText = "EXPLAIN PLAN FOR " + queryText;
+    return queryPlan(format);
+  }
+
+  private QuerySummary produceSummary(BufferingQueryEventListener listener) {
+    long start = System.currentTimeMillis();
+    int recordCount = 0;
+    int batchCount = 0;
+    QueryId queryId = null;
+    loop:
+    for ( ; ; ) {
+      QueryEvent event = listener.get();
```

Comment:
Really? The listener handles errors as follows:
```
  @Override
  public void submissionFailed(UserException ex) {
    silentPut( new QueryEvent( ex ) );
  }
```

That is, when an error arrives from the server, we create an event and put it into the event queue, later to be retrieved by the above get. Perhaps show me the error case you discovered as I'm likely missing something here.

I've seen other tests where any failure is delivered using the `submissionFailed` call...